packages/components/src/z-stack/component.tsx

```diff
@@ -0,0 +1,108 @@
+/**
+ * External dependencies
+ */
+import { css, cx } from 'emotion';
+// eslint-disable-next-line no-restricted-imports
+import type { Ref, ReactNode } from 'react';
+
+/**
+ * WordPress dependencies
+ */
+import { isValidElement } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { getValidChildren } from '../ui/utils/get-valid-children';
+import { contextConnect, useContextSystem } from '../ui/context';
+// eslint-disable-next-line no-duplicate-imports
+import type { ViewOwnProps } from '../ui/context';
+import { View } from '../view';
+import * as styles from './styles';
+const { ZStackView } = styles;
+
+export interface ZStackProps {
+	/**
+	 * Layers children elements on top of each other (first: highest z-index, last: lowest z-index).
+	 *
+	 * @default true
+	 */
+	isLayered?: boolean;
+	/**
+	 * Reverse the layer ordering (first: lowest z-index, last: highest z-index).
+	 *
+	 * @default false
+	 */
+	isReversed?: boolean;
+	/**
+	 * The amount of space between each child element.
+	 *
+	 * @default 0
+	 */
+	offset?: number;
+	/**
+	 * Child elements.
+	 */
+	children: ReactNode;
+}
+
+function ZStack(
+	props: ViewOwnProps< ZStackProps, 'div' >,
+	forwardedRef: Ref< any >
+) {
+	const {
+		children,
+		className,
+		isLayered = true,
+		isReversed = false,
+		offset = 0,
+		...otherProps
+	} = useContextSystem( props, 'ZStack' );
+
+	const validChildren = getValidChildren( children );
+	const childrenCount = validChildren.length - 1;
+
+	const clonedChildren = validChildren.map( ( child, index ) => {
+		const zIndex = isReversed ? childrenCount - index : index;
+
+		const classes = cx(
+			isLayered ? styles.positionAbsolute : styles.positionRelative,
+			css( {
+				marginLeft: ! isLayered && offset * -1,
```

Comment:
It feels a bit counter-intuitive to me that higher (positive) values of the `offset` property bring the stacked items closer together, and lower (negative) values push the items further apart

It also looks like the `offset` property does not affect the children when `isLayered: true`. I would personally expect it to work regardless of the elements being layered or not

Reply:
Do you have a better name than offset? Or should the offset start from like... all of the items completely layered together and then as you increase offset it spreads them out? I'll have to think about how to make it work 🤔 It seems like `isLayered` should just be the default, and maybe shouldn't exist as an offset of `0` in that case would make it be "layered" I'll fiddle with this a little bit.

Reply:
I think what's difficult about this is that there's no clear way to make the offset just "the width of the item" unless you happent to know the width of the item. Maybe renaming `offset` to `overlap` would make more sense?

Reply:
I took some time to tinker around and potentially came up with a way that doesn't rely on the item's width:

- kept the name for the `offset` property, but change the way it behaves. An `offset` of 0 means "no changes to how the layout flow". A positive value means that the children are further apart from each other.  A negative value means that the children are closer to each other
- made the `offset` affect the layout also when the `isLayered` and/or `isReversed` properties are `true`
- as a consequence, removed the "hover" effect from the Storybook example, as in my opinion it was making it confusing to understand how the offset works (I find that using the knobs is a more clear way)

This is the diff of the changes that I came up with. Happy to discuss this more in details in a call!

```diff
diff --git a/packages/components/src/z-stack/component.tsx b/packages/components/src/z-stack/component.tsx
index fd9b90ec52..e38c8035e1 100644
--- a/packages/components/src/z-stack/component.tsx
+++ b/packages/components/src/z-stack/component.tsx
@@ -35,11 +35,11 @@ export interface ZStackProps {
 	 */
 	isReversed?: boolean;
 	/**
-	 * The amount of overlap between each child element.
+	 * The amount of offset between each child element.
 	 *
 	 * @default 0
 	 */
-	overlap?: number;
+	offset?: number;
 	/**
 	 * Child elements.
 	 */
@@ -55,20 +55,23 @@ function ZStack(
 		className,
 		isLayered = true,
 		isReversed = false,
-		overlap = 0,
+		offset = 0,
 		...otherProps
 	} = useContextSystem( props, 'ZStack' );
 
 	const validChildren = getValidChildren( children );
-	const childrenCount = validChildren.length - 1;
+	const childrenLastIndex = validChildren.length - 1;
 
 	const clonedChildren = validChildren.map( ( child, index ) => {
-		const zIndex = isReversed ? childrenCount - index : index;
+		const zIndex = isReversed ? childrenLastIndex - index : index;
+		const offsetAmount = offset * index;
 
 		const classes = cx(
 			isLayered ? styles.positionAbsolute : styles.positionRelative,
 			css( {
-				marginLeft: ( ! isLayered && overlap * -1 ) || undefined,
+				...( isLayered
+					? { marginLeft: offsetAmount }
+					: { right: offsetAmount * -1 } ),
 			} )
 		);
 
@@ -87,17 +90,10 @@ function ZStack(
 		);
 	} );
 
-	const classes = cx(
-		css( {
-			paddingLeft: ! isLayered ? overlap : undefined,
-		} ),
-		className
-	);
-
 	return (
 		<ZStackView
 			{ ...otherProps }
-			className={ classes }
+			className={ className }
 			ref={ forwardedRef }
 		>
 			{ clonedChildren }
diff --git a/packages/components/src/z-stack/stories/index.js b/packages/components/src/z-stack/stories/index.js
index 31d707d5db..52a4549035 100644
--- a/packages/components/src/z-stack/stories/index.js
+++ b/packages/components/src/z-stack/stories/index.js
@@ -1,7 +1,3 @@
-/**
- * WordPress dependencies
- */
-import { useState } from '@wordpress/element';
 /**
  * External dependencies
  */
@@ -42,20 +38,15 @@ const Avatar = ( { backgroundColor } ) => {
 };
 
 const AnimatedAvatars = () => {
-	const [ isHover, setIsHover ] = useState( false );
-	const overlap = number( 'overlap', 20 );
 	const props = {
-		overlap: isHover ? 0 : overlap,
+		offset: number( 'offset', -20 ),
 		isLayered: boolean( 'isLayered', false ),
 		isReversed: boolean( 'isReversed', false ),
 	};
 
 	return (
 		<HStack>
-			<View
-				onMouseLeave={ () => setIsHover( false ) }
-				onMouseEnter={ () => setIsHover( true ) }
-			>
+			<View>
 				<ZStack { ...props }>
 					<Avatar backgroundColor="#444" />
 					<Avatar backgroundColor="#777" />

```

Reply:
As agreed, I committed the changes suggested above (plus changing the default value of `isLayered` to `true`) in https://github.com/WordPress/gutenberg/pull/31613/commits/045d3706341202deb2cd119a26c48b1a7f8cc234